google/cloud/terminate_handler_test.cc

```diff
@@ -0,0 +1,62 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/terminate_handler.h"
+#include <gtest/gtest.h>
+#include <iostream>
+
+using namespace google::cloud;
+
+namespace {
+const std::string handler_msg = "Custom handler invoked. Extra description: ";
+
+void CustomHandler(const char* msg) {
+  std::cerr << handler_msg << msg << std::endl;
+  abort();
+}
+
+void CustomHandlerOld(const char* msg) { abort(); }
+}  // namespace
+
+TEST(TerminateHandler, UnsetTerminates) {
+  GetTerminateHandler();
+  EXPECT_DEATH_IF_SUPPORTED(Terminate("Test"),
+                            "Aborting because exceptions are disabled: Test");
+}
+
+TEST(TerminateHandler, SettingGettingWorks) {
+  SetTerminateHandler(&CustomHandler);
+  TerminateHandler set_handler = GetTerminateHandler();
+  ASSERT_EQ(CustomHandler, *set_handler.target<void (*)(const char*)>());
```

Comment:
I think using `ASSERT_*` will require types that can be printed, you may have to do something like:

```C++
ASSERT_TRUE(CustomHandler == ...) << "The handler objects should be equal";
```

Reply:
Fascinating, functions are printable in gcc, no idea why MSVC doesn't like them, but probably not worth the investigation.